DOC: add note on type casting to numpy.left_shift().

flaviu-gostin opened the pull request:
It is not obvious how numpy.left_shift() infers types and this may lead
to unexpected outputs.  Added note to draw attention to the importance
of input types.  Added three examples.  "Closes #12310"

<!-- Please be sure you are following the instructions in the dev guidelines
http://www.numpy.org/devdocs/dev/development_workflow.html
-->

<!-- We'd appreciate it if your commit message is properly formatted
http://www.numpy.org/devdocs/dev/development_workflow.html#writing-the-commit-message
-->

flaviu-gostin committed:
DOC: add note on type casting to numpy.left_shift().

It is not obvious how numpy.left_shift() infers types and this may lead
to unexpected outputs.  Added note to draw attention to the importance
of input types.  Added three examples.  "Closes #12310"

seberg commented on numpy/core/code_generators/ufunc_docstrings.py:
```diff
@@ -1835,6 +1835,18 @@ def add_newdoc(place, name, doc):
+    Note: As explained in :ref:`Casting Rules <ufuncs.casting>`, if a ufunc
+    does not have a core loop implementation for the input types provided,
+    it will cast one or more inputs to a different type.  This may lead to
+    an unexpected result in some cases.
```
I think part of what causes the confusion here is also the fact that the second arguments dtype feels like it should not have any effect, but it does.  On the implementation side you are completely right, but I am curious if we can phrase it that way and if that may be easier to understand?

flaviu-gostin replied:
I agree.  It should be made easier to understand and the link to the Casting Rules points people in the right direction if they want to go deeper.  I modified the note accordingly.  Let me know how it sounds.

flaviu-gostin committed:
DOC: add note on type casting to numpy.left_shift(). (gh-12310)

Intuitively, one expects the dtype of the first argument in
np.left_shift() to be preserved and assigned to the output.  However,
this may be changed by the dtype of the second argument and lead to
confusing results.  Added note to draw attention to this.  Added three
examples.  Closes gh-12310.

seberg commented on numpy/core/code_generators/ufunc_docstrings.py:
```diff
@@ -1835,6 +1835,17 @@ def add_newdoc(place, name, doc):
     >>> np.left_shift(5, [1,2,3])
     array([10, 20, 40])
 
+    Note: The dtype of the second argument may change the dtype of the first
+    argument (see :ref:`Casting Rules <ufuncs.casting>`) and lead to unexpected
```
```suggestion
    Note that the dtype of the second argument may change the dtype of the
    result and can lead to unexpected results in some cases
    (see :ref:`Casting Rules <ufuncs.casting>`):
```
Thanks for following up! Some thoughts only, but I like this either way. The only real suggestion is to change it to the `dtype of the result` (or output) instead of `first argument`.

flaviu-gostin replied:
That sounds good.  Thanks!  I made the changes you suggested.

flaviu-gostin committed:
Made change suggested by seberg in PR

rossbar commented on numpy/core/code_generators/ufunc_docstrings.py:
```diff
@@ -1835,6 +1835,17 @@ def add_newdoc(place, name, doc):
+    array([254], dtype=uint8)
+    >>> np.left_shift(np.array([255], dtype='u1'), np.array([1], dtype='i1'))
+    array([510], dtype=int16)
+    >>> np.left_shift(np.uint8(255), 1)
```
Perhaps you want to lead with the scalar example? It might be helpful to have a comment that gives the expected result (254) .

flaviu-gostin replied:
Yes, those examples were not conveying a very clear picture.  It should be clearer now.  I removed the array examples to keep it simple and clear.

rossbar reviewed (commented):
The text looks good to me, just a comment about the ordering of the code examples and maybe adding a clarifying comment.

flaviu-gostin committed:
Made changes as suggested by rossbar in PR

Used the examples of zhanwenchen

rossbar commented on numpy/core/code_generators/ufunc_docstrings.py:
```diff
@@ -1835,6 +1835,15 @@ def add_newdoc(place, name, doc):
+    result and can lead to unexpected results in some cases (see
+    :ref:`Casting Rules <ufuncs.casting>`):
+
+    >>> a = np.left_shift(np.uint8(255), np.uint8(1)); print(a, type(a))
+    254 <class 'numpy.uint8'> # as expected
+    >>> b = np.left_shift(np.uint8(255), 1); print(b, type(b))
+    510 <class 'numpy.int64'> # unexpected
```
```suggestion
    >>> a = np.left_shift(np.uint8(255), 1) # Expect 254
    >>> print(a, type(a)) # Unexpected result due to upcasting
    510 <class 'numpy.int64'>
    >>> b = np.left_shift(np.uint8(255), np.uint8(1))
    >>> print(b, type(b))
    254 <class 'numpy.uint8'>

```

rossbar reviewed (commented):
A couple further suggestions: I think having the negative example first is more compelling as it illustrates the problem, followed by an example that shows how being explicit about the types of *both* inputs avoids the casting problem.

I also suggest un-chaining the expressions (i.e. putting the `print`s on a separate line) as I think it's more clear

flaviu-gostin committed:
Rearrange examples in note on type casting on ufunc left_shift

Changes suggested by rossbar in PR.
Start with the negative example.  Also, unchain expressions.

Co-Authored-By: Ross Barnowski <rossbar@berkeley.edu>

rossbar reviewed (approved):
LGTM, the removal of the chained expressions in the examples is an improvement.

seberg commented:
Hmmm, universal test failure on windows there (not related to this). Which makes me realize that `numpy.int64` is not quite right on windows (or 32bit linux) in principle. IIRC we do not really check this for the tests, and I guess it does not matter much here?

flaviu-gostin commented:
> LGTM, the removal of the chained expressions in the examples is an improvement.

Thanks for that.  It looks better like this.

rossbar commented:
Good point @seberg , I hadn't thought of that. Perhaps add a `# May vary`?

flaviu-gostin commented:
Apologies for the naive question: why are there test failures if the only changes are on the documentation?
I see that @WarrenWeckesser merged. Did you find out what the problem was on that branch in scipy?
Should I rebase? Would it help?

WarrenWeckesser commented:
Closing and reopening to re-run tests.

rossbar commented:
@flaviu-gostin sorry for all the confusion - something went awry with the NumPy CI in the past couple days: @WarrenWeckesser linked PR above is about that issue and not related to the content of this PR. 

The azure pipeline issue has been cleared up. No rebase is required - @WarrenWeckesser 's "power cycling" of the PR has re-triggered the CI, so we're all good. You're all set here - we'll wait to see if any other reviewers have comments but your changes LGTM!

mattip commented:
Thanks @flaviu-gostin
